src/diffusers/pipelines/unidiffuser/modeling_text_decoder.py

```diff
@@ -0,0 +1,188 @@
+from typing import Optional
+
+import numpy as np
+import torch
+from torch import nn
+from transformers import GPT2LMHeadModel, GPT2Tokenizer
+
+from ...configuration_utils import ConfigMixin, register_to_config
+from ...models import ModelMixin
+
+
+# Modified from ClipCaptionModel in https://github.com/thu-ml/unidiffuser/blob/main/libs/caption_decoder.py
+class UniDiffuserTextDecoder(ModelMixin, ConfigMixin):
+    @register_to_config
+    def __init__(
+        self,
+        tokenizer: GPT2Tokenizer,
+        text_decoder: GPT2LMHeadModel,
+        prefix_length: int,
+        hidden_dim: Optional[int] = None,
+        use_hidden_dim: bool = True,
+    ):
+        """
+        Text decoder model for a image-text [UniDiffuser](https://arxiv.org/pdf/2303.06555.pdf) model. This is used to
+        generate text from the UniDiffuser image-text embedding.
+
+        Parameters:
+            tokenizer ([`GPT2Tokenizer`]):
+                Tokenizer of class
+                [GPT2Tokenizer](https://huggingface.co/docs/transformers/model_doc/gpt2#transformers.GPT2Tokenizer) for
+                the GPT-like text decoder model.
+            text_decoder ([`GPT2LMHeadModel`]):
+                Text decoder model of class
+                [GPT2LMHeadModel](https://huggingface.co/docs/transformers/model_doc/gpt2#transformers.GPT2LMHeadModel)
+                used to generate text from the UniDiffuser text embedding.
+            prefix_length (`int`):
+                TODO
+            hidden_dim (`int`, *optional*):
+                Hidden dim of the MLP if we encode the prefix.
+            use_hidden_dim (`bool`, *optional*, defaults to `True`):
+                Whether or not to use a MLP to encode the prefix.
+        """
+        super().__init__()
+        self.prefix_length = prefix_length
+
+        eos = "<|EOS|>"
+        special_tokens_dict = {"eos_token": eos}
+        self.tokenizer = tokenizer
+        self.tokenizer.add_special_tokens(special_tokens_dict)
```

Comment:
Note that we can do this directly for the uploaded tokenizer. E.g. let's just upload a tokenizer that has EOS already added so that we don't have to do it every time we call the model at __init__ 

More than happy to help here later on!

Reply:
Removed the tokenizer logic from `__init__`, will work on uploading the appropriate tokenizer.

Reply:
I've prepared some native diffusers checkpoints for the current implementation of the `UniDiffuserPipeline` and its building blocks (e.g. `UniDiffuserModel`, `UniDiffuserTextDecoder`, etc.) [see the `convert_to_ckpt.py` script]. How can I upload these up to the hub?

Reply:
I was able to upload some models to the hub (see e.g. small test models [here](https://huggingface.co/dg845/unidiffuser-diffusers-test/tree/main)), but I'm confused about how to save/push to hub a tokenizer with added special tokens. The [documentation](https://huggingface.co/docs/transformers/v4.27.2/en/internal/tokenization_utils#transformers.PreTrainedTokenizerBase.save_pretrained) for `PreTrainedTokenizerBase.from_pretrained` says that it won't save modifications to the tokenizer after initialization and I wasn't able to find any resources on how to do it after searching.

For reference, the [code in the base unidiffuser library](https://github.com/thu-ml/unidiffuser/blob/688f7b276396ac05931e8b014f2bc5e43239f3fd/libs/caption_decoder.py#L57-L62) is something like

```python
eos = '<|EOS|>'
special_tokens_dict = {'eos_token': eos}
base_tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
base_tokenizer.add_special_tokens(special_tokens_dict)
```